libraries/cyclestreets-view/src/main/java/net/cyclestreets/routing/RoutingTask.kt

```diff
@@ -2,49 +2,84 @@ package net.cyclestreets.routing
 
 import android.content.Context
 import android.os.AsyncTask
+import android.util.Log
 import android.widget.Toast
+import net.cyclestreets.CycleStreetsPreferences
+import net.cyclestreets.RoutePlans.PLAN_LEISURE
 
 import net.cyclestreets.api.JourneyPlanner
 import net.cyclestreets.content.RouteData
 import net.cyclestreets.util.Dialog
+import net.cyclestreets.util.Logging
 import net.cyclestreets.util.ProgressDialog
+import net.cyclestreets.view.R
 
 abstract class RoutingTask<Params> protected constructor(private val initialMsg: String,
                                                          private val context: Context) : AsyncTask<Params, Int, RouteData>() {
     private var progress: ProgressDialog? = null
     private var error: String? = null
+    private val TAG = Logging.getTag(RoutingTask::class.java)
 
     protected constructor(progressMessageId: Int,
                           context: Context) : this(context.getString(progressMessageId), context)
 
-    protected fun fetchRoute(routeType: String,
-                             speed: Int,
-                             waypoints: Waypoints): RouteData? {
-        return fetchRoute(routeType, -1, speed, waypoints)
-    }
-
-    @JvmOverloads
     protected fun fetchRoute(routeType: String,
                              itinerary: Long,
                              speed: Int,
-                             waypoints: Waypoints? = null): RouteData? {
+                             waypoints: Waypoints? = null,
+                             distance: Int? = null,
+                             duration: Int? = null,
+                             pois: String? = null): RouteData? {
         return try {
-            val json = doFetchRoute(routeType, itinerary, speed, waypoints)
-            RouteData(json, waypoints, null)
+            val json = doFetchRoute(routeType, itinerary, speed, waypoints, distance, duration, pois)
+
+            when {
+                (json == null) -> {
+                    error = context.getString(R.string.route_not_found)
+                    throw Exception(error)
+                }
+                json.contains("Error", ignoreCase = true) -> {
+                    // Put space after ":", remove quotes, drop "{" at start, drop "}" at end
+                    //e.g. {"Error":"The requested itinerary cannot be found."} -> Error: The requested itinerary cannot be found.
+                    error = json.replace(":\"", ": ")
+                                .replace(""""""", "", true)
+                                .drop(1)
```

Comment:
I absolutely cannot accept manual parsing of a JSON string, and neither should you 😉 

Suggest you instantiate an `ObjectMapper` at appropriate scope (the class?) and then either:
- define a POJO representing the JSON model, and parse into that, or:
- parse into ObjectNode, and then read the value of the `error` field, as explained in https://stackoverflow.com/a/26191723/2108057
`objectMapper.readValue(json, JourneyDomainObject::class.java)

Reply:
Yes, never parse JSON manually is a lesson we learnt :)

Reply:
My rule of thumb is always to ask myself ten times a day "am I doing something tedious which surely must be written better already in a library?"

Reply:
Using JsonNode seemed the simplest option for a single field.
Having "Error: " in the toast message seems superfluous, so I left that out of the toast message and just display the message itself.
<img width="183" alt="Error message from server as toast" src="https://user-images.githubusercontent.com/59799824/111033009-047efa00-8407-11eb-9432-fef8b11ddc5e.PNG">